src/libexpr/primops.cc

```diff
@@ -83,22 +83,21 @@ StringMap EvalState::realiseContext(const NixStringContext & context)
     for (auto & d : drvs) buildReqs.emplace_back(DerivedPath { d });
     store->buildPaths(buildReqs);
 
-    /* Get all the output paths corresponding to the placeholders we had */
     for (auto & drv : drvs) {
         auto outputs = resolveDerivedPath(*store, drv);
         for (auto & [outputName, outputPath] : outputs) {
-            res.insert_or_assign(
-                DownstreamPlaceholder::unknownCaOutput(drv.drvPath, outputName).render(),
-                store->printStorePath(outputPath)
-            );
-        }
-    }
-
-    /* Add the output of this derivations to the allowed
-       paths. */
-    if (allowedPaths) {
-        for (auto & [_placeholder, outputPath] : res) {
-            allowPath(store->toRealPath(outputPath));
+            /* Add the output of this derivations to the allowed
+               paths. */
+            if (allowedPaths) {
+                allowPath(outputPath);
+            }
+            /* Get all the output paths corresponding to the placeholders we had */
+            if (experimentalFeatureSettings.isEnabled(Xp::CaDerivations)) {
+                res.insert_or_assign(
+                    DownstreamPlaceholder::unknownCaOutput(drv.drvPath, outputName).render(),
+                    store->printStorePath(outputPath)
+                );
+            }
```

Comment:
Shouldn't there be an "else" here (like throwing an error)? Seems strange that we can do nothing if CA derivations are not enabled.

Reply:
I don't think so actually. It is unclear but basically:

1. The only way to get these placeholders is the `outPath`s of CA derivation outputs.
2. If we don't have those, then there is nothing to substitute.
3. If someone tried to hand-write one of these (rather than using `outPath`, it shouldn't work.

Today, it does work; this is mistakenly a bonus feature that can be used without `ca-derivations` with careful hand-coding. But I don't think we want that.